File: packages/faro-ios-sdk/Sources/FaroSDK/Internal/InternalLogger.swift
import Foundation
import os.log

public enum InternalLoggerLevel: Int, Comparable {
    case verbose = 0
    case debug = 1
    case info = 2
    case warn = 3
    case error = 4
    case none = 5

    public static func < (lhs: InternalLoggerLevel, rhs: InternalLoggerLevel) -> Bool {
        return lhs.rawValue < rhs.rawValue
    }
}

internal final class InternalLogger {
    private let level: InternalLoggerLevel
    private let logger = OSLog(subsystem: "com.grafana.faro", category: "Faro")

    init(level: InternalLoggerLevel = .error) {
        self.level = level
    }

    func verbose(_ message: String) {
        guard level <= .verbose else { return }
        os_log(.debug, log: logger, "%{public}@", message)
    }

    func debug(_ message: String) {
        guard level <= .debug else { return }
        os_log(.debug, log: logger, "%{public}@", message)
    }

    func info(_ message: String) {
        guard level <= .info else { return }
        os_log(.info, log: logger, "%{public}@", message)
    }

    func warn(_ message: String) {
        guard level <= .warn else { return }
        os_log(.default, log: logger, "WARN: %{public}@", message)
    }

    func error(_ message: String, error: Error? = nil) {
        guard level <= .error else { return }
        if let error = error {
            os_log(.error, log: logger, "%{public}@: %{public}@", message, error.localizedDescription)
        } else {
            os_log(.error, log: logger, "%{public}@", message)
        }
    }
}

File: packages/faro-android-sdk/src/main/kotlin/com/grafana/faro/internal/InternalLogger.kt
package com.grafana.faro.internal

import android.util.Log

enum class InternalLoggerLevel {
    VERBOSE, DEBUG, INFO, WARN, ERROR, NONE
}

internal class InternalLogger(
    private val level: InternalLoggerLevel = InternalLoggerLevel.ERROR
) {
    companion object {
        private const val TAG = "Faro"
    }

    fun verbose(message: String) {
        if (level <= InternalLoggerLevel.VERBOSE) Log.v(TAG, message)
    }

    fun debug(message: String) {
        if (level <= InternalLoggerLevel.DEBUG) Log.d(TAG, message)
    }

    fun info(message: String) {
        if (level <= InternalLoggerLevel.INFO) Log.i(TAG, message)
    }

    fun warn(message: String) {
        if (level <= InternalLoggerLevel.WARN) Log.w(TAG, message)
    }

    fun error(message: String, throwable: Throwable? = null) {
        if (level <= InternalLoggerLevel.ERROR) Log.e(TAG, message, throwable)
    }
}
